src/ParallelAlgorithms/NonlinearSolver/NewtonRaphson/ElementActions.hpp

```diff
@@ -0,0 +1,508 @@
+// preparation. We only need to set the initial guess.
+//
+// We also prepare the line-search globalization here. Since we don't know if
+// a step will be sufficient before taking it, we have to store the original
+// field values.
+//
+// The algorithm jumps back to this action from `CompleteStep` to continue
+// iterating.
+template <typename FieldsTag, typename OptionsGroup, typename Label>
+struct PrepareStep {
+ private:
+  using fields_tag = FieldsTag;
+  using correction_tag =
+      db::add_tag_prefix<NonlinearSolver::Tags::Correction, fields_tag>;
+  using linear_operator_applied_to_correction_tag =
+      db::add_tag_prefix<LinearSolver::Tags::OperatorAppliedTo, correction_tag>;
+  using globalization_fields_tag =
+      db::add_tag_prefix<NonlinearSolver::Tags::Globalization, fields_tag>;
+
+ public:
+  using const_global_cache_tags =
+      tmpl::list<NonlinearSolver::Tags::DampingFactor<OptionsGroup>,
+                 logging::Tags::Verbosity<OptionsGroup>>;
+
+  template <typename DbTagsList, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static std::tuple<db::DataBox<DbTagsList>&&> apply(
+      db::DataBox<DbTagsList>& box,
+      const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      const Parallel::GlobalCache<Metavariables>& /*cache*/,
+      const ArrayIndex& array_index, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    if (UNLIKELY(get<logging::Tags::Verbosity<OptionsGroup>>(box) >=
+                 ::Verbosity::Debug)) {
+      Parallel::printf(
+          "%s " + Options::name<OptionsGroup>() + "(%zu): Prepare step\n",
+          array_index,
+          db::get<Convergence::Tags::IterationId<OptionsGroup>>(box) + 1);
+    }
+
+    db::mutate<Convergence::Tags::IterationId<OptionsGroup>, correction_tag,
+               linear_operator_applied_to_correction_tag,
+               NonlinearSolver::Tags::Globalization<
+                   Convergence::Tags::IterationId<OptionsGroup>>,
+               NonlinearSolver::Tags::StepLength<OptionsGroup>,
+               globalization_fields_tag>(
+        make_not_null(&box),
+        [](const gsl::not_null<size_t*> iteration_id, const auto correction,
+           const auto linear_operator_applied_to_correction,
+           const gsl::not_null<size_t*> globalization_iteration_id,
+           const gsl::not_null<double*> step_length,
+           const auto globalization_fields, const auto& fields,
+           const double damping_factor) noexcept {
+          ++(*iteration_id);
+          // Begin the linear solve with a zero initial guess
+          *correction =
+              make_with_value<typename correction_tag::type>(fields, 0.);
+          // Since the initial guess is zero, we don't need to apply the linear
+          // operator to it but can just set it to zero as well. Linear things
+          // are nice :)
```

Comment:
It's sort of amusing how much of numerical work is taking non-linear things and finding better and better linear (or sequences of linear) approximations to them :P

Reply:
That's because linear things _are_ nice :)

Actually there exists also a whole class of nonlinear solvers that don't use linear approximations, see here in case you are interested: https://www.mcs.anl.gov/petsc/documentation/nonlinearsolvertable.html. They are basically nonlinear variants of linear solvers. However the Newton-Krylov solver (the one I'm implementing here, which runs a linear solve in each iteration) has the advantage that we can apply all of the fancy techniques for efficiently solving linear systems, including all the preconditioning methods.